Add tests for leg cit

georgd opened the pull request.

georgd committed:
Ignore Mac specific files and Vim swap files

georgd committed:
Initial commit

georgd committed:
More leg cit tests

fbennett commented:
Looking good!

Running these against your latest checkin of "JM leg cit 3 mit Rechtsquellenverzeichnis und Literaturverzeichnis," I'm getting errors on five fixtures. Two seem to be correct, but fail for a lowercase OGH (if you update with `npm update citeproc-test-runner` to pull in legal abbreviation support and a working copy of the AT abbrevs, I think you'll get them as well - the RESULT section in the tests currently uses the raw court code).

If you get stuck with remaining issues, just give a shout.

fbennett commented:
Re the priority of the various short-forms (from mail), if the current behavior works for your requirements and aligns with Zotero, I'll keep things exactly as they are.

fbennett commented:
Re this test (from mail):

https://github.com/georgd/jm-style-tests/blob/master/jm-leg-cit-3-rechtsquellenverzeichnis-literaturverzeichnis/style_test004.txt

The test is passing for me here.

fbennett commented:
Re this test (from mail):

https://github.com/georgd/jm-style-tests/blob/master/jm-leg-cit-3-rechtsquellenverzeichnis-literaturverzeichnis/style_test027.txt

There was a bug in the latest processor release, which I've now fixed. Update the test runner with "npm update citeproc-test-runner" for the latest (v1.1.9).

In the test fixture, some small amendments are needed to get it working. A patch is attached:

[LEG-CIT-027.zip](https://github.com/Juris-M/jm-style-tests/files/3564029/LEG-CIT-027.zip)

In the running test, `references` is not being pulled in because the `jurism` macro in the style does not call the module macro that contains it (`juris-main-short`). That leads to a larger structural/design issue that we should consider at this point.

The ultimate aim is for users to be able to cite material from arbitrary jurisdictions, and for the system to cite each resource by its local rules, adapted to the parent style's general guidelines. For a basic example, in Bluebook style, an English and a US case, in first-reference and back-reference form, might look like this:

* Donoghue v. Stevenson, (1932) S.C. 31 (HL).
  *Donoghue*, (1932) S.C. 31.
* Escola v. Coca Cola Bottling Co., 24 Cal. 2d 453 (1944).
  *Escola*, 24 Cal. 2d 453.

In the OSCOLA style, the same references would be styled like this:

* *Donoghue v Stevenson* (1932) SC 31 (HL).
  *Donoghue*.
* *Escola v. Coca Cola Bottling Co.* 24 Cal. 2d 453 (1944).
  *Escola*.

The two parent style variants use the same style module, but select and format the elements differently in their respective "jurism" macros. You can get Common Law citation forms for free in the "JM Leg Cit" style by taking the "jurism" macro of one of them as a starting point (I would suggest OSCOLA as the better style), and extending it for use with Austrian material. Then we can fold your own "jurism" macro back into the other styles, and also get the Austrian family of styles for free in Bluebook, OSCOLA, etc.

Case citations are the hardest to generalize to a "jurism" meta-pattern. The Austrian references seem to place the locator at the opposite end of the citation, so it looks like the US/UK/Commonwealth pattern in OSCOLA "jurism" is indeed unsuitable, and casting entirely separate CSL code, as you are doing in JM Leg Cit, is the right thing to do.

When we reach the point of combining the two patterns in a single general-purpose "jurism" macro, we'll need to settle on a means of distinguishing "Austria family" items from "Common Law family" items. Depending on how much variance there is in the general pattern of Continental case citations, there are at least two options:
1. Treat all case items without a title as continental-style references; or
2. Select "jurism" patterns by country.

We can settle on the approach later, but the more general the "jurism" pattern can be made within the family of civil law citation practices, the lighter the maintenance burdens will be down the road.

georgd commented:
> Re this test (from mail):
> 
> https://github.com/georgd/jm-style-tests/blob/master/jm-leg-cit-3-rechtsquellenverzeichnis-literaturverzeichnis/style_test004.txt
> 
> The test is passing for me here.

Yes, it’s passing here as well but when I cite this entry using Juris-M, it won’t let me enter an abbreviation for the title (which would overwrite the short title if used with `title form="short"`.

georgd commented:
> Looking good!
> 
> Running these against your latest checkin of "JM leg cit 3 mit Rechtsquellenverzeichnis und Literaturverzeichnis," I'm getting errors on five fixtures. Two seem to be correct, but fail for a lowercase OGH (if you update with `npm update citeproc-test-runner` to pull in legal abbreviation support and a working copy of the AT abbrevs, I think you'll get them as well - the RESULT section in the tests currently uses the raw court code).

I updated citeproc-test-runner but still no luck.

fbennett commented:
> Yes, it’s passing here as well but when I cite this entry using Juris-M, it won’t let me enter an abbreviation for the title (which would overwrite the short title if used with `title form="short"`.

Oops, sorry, I misunderstood. Will look again.

fbennett commented:
> I updated citeproc-test-runner but still no luck.

Backing up a bit, I should ask how you have installed the test runner. If by cloning the source repo, you would enter the repo, do "git pull," then "npm update." If it was installed via NPM, you would just do "npm update citeproc-test-runner." If the test is expecting lowercase (the court ID), it will fail when the processor delivers the abbreviation (uppercase). Are you still getting lowercase, unabbreviated output?

georgd committed:
Add title to Austrian legal_case

georgd committed:
Correct the expected output

georgd committed:
Correct fixture for disambiguation test

fbennett commented:
> Yes, it’s passing here as well but when I cite this entry using Juris-M, it won’t let me enter an abbreviation for the title (which would overwrite the short title if used with `title form="short"`.

On this item, abbreviations are offered when a style calls title with form="short". It initially failed for me when I ran the word-processor, because my `style-modules` folder for testing is in one location, and my live Jurism data folder is in another. That may be happening for you as well -- there is an AT module, but it only supplies `legal_case`, so Jurism would fall back to US, which does not abbreviate the title in `jurism-title-short`.

Separately, is there are reason why your "jurism" macro calls `jurism-title-short` always, and never `jurism-title`? Ordinarily the latter would be used for position="first", and the former for position="subsequent". If there is no difference between the two, it's safe to call one macro from the other inside the module, to save the trouble of updating both of them.

georgd commented:
> > I updated citeproc-test-runner but still no luck.
> 
> Backing up a bit, I should ask how you have installed the test runner. If by cloning the source repo, you would enter the repo, do "git pull," then "npm update." If it was installed via NPM, you would just do "npm update citeproc-test-runner." If the test is expecting lowercase (the court ID), it will fail when the processor delivers the abbreviation (uppercase). Are you still getting lowercase, unabbreviated output?

I updated it using "npm update citeproc-test-runner".

Im still getting lowercase ogh in place of the expected OGH:
      -  ogh 4 Ob 8/11x.
      +  OGH 4 Ob 8/11x.

georgd commented:
> > Yes, it’s passing here as well but when I cite this entry using Juris-M, it won’t let me enter an abbreviation for the title (which would overwrite the short title if used with `title form="short"`.
> 
> On this item, abbreviations are offered when a style calls title with form="short". It initially failed for me when I ran the word-processor, because my `style-modules` folder for testing is in one location, and my live Jurism data folder is in another. That may be happening for you as well -- there is an AT module, but it only supplies `legal_case`, so Jurism would fall back to US, which does not abbreviate the title in `jurism-title-short`.

Ok, that did the trick :-)

> Separately, is there are reason why your "jurism" macro calls `jurism-title-short` always, and never `jurism-title`? Ordinarily the latter would be used for position="first", and the former for position="subsequent". If there is no difference between the two, it's safe to call one macro from the other inside the module, to save the trouble of updating both of them.

It calls jurism-title for the bibliography and jurism-title-short for all citations. Please bear in mind that I’ll have to add a second style which will not create a bibliography where the first citation will be formatted (almost) exactly like the bibliography entries. In the style which I’m currently working on, all citations are like subsequent citations in as short as possible form.

georgd commented:
> Re this test (from mail):
> 
> https://github.com/georgd/jm-style-tests/blob/master/jm-leg-cit-3-rechtsquellenverzeichnis-literaturverzeichnis/style_test027.txt
> 
> There was a bug in the latest processor release, which I've now fixed. Update the test runner with "npm update citeproc-test-runner" for the latest (v1.1.9).
> 
> In the test fixture, some small amendments are needed to get it working. A patch is attached:
> 
> [LEG-CIT-027.zip](https://github.com/Juris-M/jm-style-tests/files/3564029/LEG-CIT-027.zip)
> 
> In the running test, `references` is not being pulled in because the `jurism` macro in the style does not call the module macro that contains it (`juris-main-short`). That leads to a larger structural/design issue that we should consider at this point.

For the time being (until resolving what you propose), I fixed this in the module and the style but it doesn’t seem to pick up the disambiguation information.

> The ultimate aim is for users to be able to cite material from arbitrary jurisdictions, and for the system to cite each resource by its local rules, adapted to the parent style's general guidelines. For a basic example, in Bluebook style, an English and a US case, in first-reference and back-reference form, might look like this:
> 
> * Donoghue v. Stevenson, (1932) S.C. 31 (HL).
>   _Donoghue_, (1932) S.C. 31.
> * Escola v. Coca Cola Bottling Co., 24 Cal. 2d 453 (1944).
>   _Escola_, 24 Cal. 2d 453.
> 
> In the OSCOLA style, the same references would be styled like this:
> 
> * _Donoghue v Stevenson_ (1932) SC 31 (HL).
>   _Donoghue_.
> * _Escola v. Coca Cola Bottling Co._ 24 Cal. 2d 453 (1944).
>   _Escola_.
> 
> The two parent style variants use the same style module, but select and format the elements differently in their respective "jurism" macros. You can get Common Law citation forms for free in the "JM Leg Cit" style by taking the "jurism" macro of one of them as a starting point (I would suggest OSCOLA as the better style), and extending it for use with Austrian material. Then we can fold your own "jurism" macro back into the other styles, and also get the Austrian family of styles for free in Bluebook, OSCOLA, etc.

I hoped to be able to do so but there’s at least one fundamental issue: there’s English language hard coded in the style modules. That is fine in case of US or UK legislation and jurisdiction but that’s not appropriate at all for the EU (which I admit was the only module I looked very close because that’s essential for us, whereas US and UK jurisdiction are very rarely to be cited in Austria). I’m ok to have a look at OSCOLA again to find out if I could get closer to that model. Looking at international courts, however, I fear that it’s not really usable:

It’s expected:
ICC, Pre-Trial Chamber I, _The Prosecutor v Thomas Lubanga Dyilo_, ICC-01/04–01/06, para 35 et seqq <icc-cpi.int/iccdocs/doc/doc266175.pdf>.
IGH, _Case Concerning Military and Paramilitary Activities in and against Nicaragua (Nicaragua v USA)_, Merits, Judgment, ICJ Reports 1986, 14(114).
ICJ, _Oil Platforms (Iran v USA)_, Judgment, ICJ Reports 2003, 161, Separate Opinion of Judge Simma, para 16.

> Case citations are the hardest to generalize to a "jurism" meta-pattern. The Austrian references seem to place the locator at the opposite end of the citation, so it looks like the US/UK/Commonwealth pattern in OSCOLA "jurism" is indeed unsuitable, and casting entirely separate CSL code, as you are doing in JM Leg Cit, is the right thing to do.
> 
> When we reach the point of combining the two patterns in a single general-purpose "jurism" macro, we'll need to settle on a means of distinguishing "Austria family" items from "Common Law family" items. Depending on how much variance there is in the general pattern of Continental case citations, there are at least two options:
> 
> 1. Treat all case items without a title as continental-style references; or

That’s a mis-interpretation from my incomplete test fixtures (which I corrected now). Titles are optional, yet recommended, for national jurisdiction but obligatory for EU jursidiction. I did a quick research in languages accessible to me (French, Spanish, Italian, Portuguese) which showed me that generally it’s the same in Italy, France, Spain, Portugal and Brasil (also, laws are largely cited in a similar way). Also, the order of the elements is more or less the same:
Court – date and case number [order varies, some information may be left out] – title – reporter – commenter.

The title, if present is often between date/case number and the reporter, the leg cit style being rather an exception in placing it at the end. The other very widely used Austrian legal style, which is more traditional (and which I also always have in the back of my mind when deciding on implementation details) shows the title between case number and reporter.

> 2. Select "jurism" patterns by country.

> We can settle on the approach later, but the more general the "jurism" pattern can be made within the family of civil law citation practices, the lighter the maintenance burdens will be down the road.

That’s of course to be preferred and I really want to help with it.

fbennett commented:
> For the time being (until resolving what you propose), I fixed this in the module and the style but it doesn’t seem to pick up the disambiguation information.

I figured out what the trouble is. There is no disambiguation method specified in the parent style, so the processor is bypassing the entire disambiguation engine as a performance measure.

This should be more friendly, but in the short term, you should be able to insert an unused macro that includes a `cs:choose` statement with `disambiguate="true"`. That should be enough to enable disambiguation, and wake the condition up in the processor. If that doesn't help, you would need to turn on actual disambiguation, which would affect style output. (So in that case, I should look into fixing this sooner rather than later.)

fbennett commented:
> That’s a mis-interpretation from my incomplete test fixtures

Understood. I'll leave it in your hands - just wanted to be sure the issue was in the mix.

fbennett commented:
> I hoped to be able to do so but there’s at least one fundamental issue: there’s English language hard coded in the style modules.

After things settle down in your work on the style, let's discuss this further. I think the OSCOLA/Bluebook meta-pattern will work for the international materials, possibly with a little tweaking. On the localization of terms, there are various options. For common terms, we could extend the set of terms available in CSL-M. Alternatively, for a small subset of languages it is possible to embed conditionals that test for the language of the item. It might take a couple of iterations, but I think we can thread the needle - and improved support for international materials would be a big win.

fbennett commented:
> It calls jurism-title for the bibliography and jurism-title-short for all citations.

I understand. That makes sense.

fbennett commented:
> Im still getting lowercase ogh in place of the expected OGH:

Ugh, that's not good. I brought a Mac machine home from work today, I'll try on that with a clean install of the test runner tomorrow, and see what's going wrong with this.

georgd commented:
> > I hoped to be able to do so but there’s at least one fundamental issue: there’s English language hard coded in the style modules.
> 
> After things settle down in your work on the style, let's discuss this further. I think the OSCOLA/Bluebook meta-pattern will work for the international materials, possibly with a little tweaking. On the localization of terms, there are various options. For common terms, we could extend the set of terms available in CSL-M. Alternatively, for a small subset of languages it is possible to embed conditionals that test for the language of the item. It might take a couple of iterations, but I think we can thread the needle - and improved support for international materials would be a big win.

I think so too. Speaking of language issues there’s some even bigger: when citing EU legislation, the full form is something like
"Regulation (EU) 2016/679 of the European Parliament and of the Council of 27 April 2016..."
Ideally, this would be entered into JM as
genre: Regulation, jurisdiction: EU, authorities: EP and Council, date...
and the JM takes care of creating the citation. But that only works for languages without inflections and contractions like English, Chinese or Japanese.

In German that would be:
"Verordnung 2016/679/EU des Europäischen Parlaments und des Rates vom 27.04.2016..."
where the authorities are in genetive case (nominative: Europäisches Parlament and Rat) and the article needs to be accorded as well ("des" v. nom. "das" Parlament/"der" Rat). Were it from the Commission, it would be "VO xyz _der_ Kommission vom ...".

Similarly in Italian: ... del Parlamento/del Consiglio/della Commissione

For perfect results, CSL had some linguistic/grammatical knowledge :). Else, content testing would provide some help (I understand well that this is generally rejected by the CSL developers).

fbennett commented:
> ... where the authorities are in genetive case (nominative: Europäisches Parlament and Rat) and the article needs to be accorded as well ("des" v. nom. "das" Parlament/"der" Rat). Were it from the Commission, it would be "VO xyz _der_ Kommission vom ...".

Yes, narrative citations will be ... interesting. There was a forum discussion long ago with a user from Greece in which he pressed these issues. We didn't have a good answer then, but each day is a possibility for fresh inspiration ...

georgd commented:
Language and abbreviation: style_test013 shows an actual issue that needs to be solved. The European courts definitely should not be abbreviated by their English names but in German. Can the abbreviation file be localized somehow?

georgd committed:
Correct EU case test

fbennett commented:
> Language and abbreviation: style_test013 shows an actual issue that needs to be solved. The European courts definitely should not be abbreviated by their English names but in German. Can the abbreviation file be localized somehow?

The source format for the LRR recognizes alternative abbrevs, and the script that compiles to the auto-abbreviation files will pick them up. That feature of the source file is explained here:

https://github.com/Juris-M/legal-resource-registry#alternative-languages

You would then set "de" as one of the  module preferences (i.e. "LegCit de"). Module prefs apply to both style modules and auto-abbrevs, so you could name the variants "LegCit" in the source file, but then they would be tightly bound to that one style variant).

The instructions go through setup and configuration. You'll need a clone of the LRR, and to set the abbrevs output path to "juris-abbrevs" in your Jurism data directory. New abbrevs will take effect after Jurism is restarted.

georgd commented:
That sounds great! I haven't seen that in the abbreviation files that I looked at so I assumed such feature to be lacking. Probably looking at US, UK and EU files was just not enough :)

So, I could specify multiple module variants as well which would be checked in a fallback manner? I'm asking because the second Austrian style has an irreconcilable difference with this one (order of year and issue reversed in legislation citation) for which I'd have to duplicate the whole LegCit module.

fbennett commented:
> So, I could specify multiple module variants as well which would be checked in a fallback manner?

Yes, that's right. I'm not sure in which direction the priorities work for preferences, but it will trawl until it hits a match of preference + item type support. So for the exceptional style, you would need only code for legal case, with everything else falling back to LegCit (and thence to US, if there is still no match).

georgd committed:
Try to implement parallel citations

georgd commented on jm-leg-cit-3-rechtsquellenverzeichnis-literaturverzeichnis/style_test028.txt:
```diff
@@ -0,0 +1,120 @@
+  <i>Wilhelm</i>, ecolex 1998, 692 Abs 123
+BIBLIOGRAPHY
+<div class="csl-bib-body">
+  <div class="csl-entry">OGH 12.02.1998, 2 Ob 328/97t = SZ 71/21 = JBl 1999, 54 (<i>Karollus</i>) = ecolex 1998, 692 (<i>Wilhelm</i>).</div>
```
This doesn’t work as expected. I only get the last of the parallel cites:
```<div class="csl-entry">ogh 12.02.1998, 2 Ob 328/97t = ecolex 1998, 692 (<i>Wilhelm</i>).</div>```

Even worse, in Word I get this in the bibliography (instead of the bibliography line as in the test fixture):
![image](https://user-images.githubusercontent.com/214002/64908071-923b3580-d6fb-11e9-9c6c-2f652fab5fe0.png)
and this in the footnote:
![image](https://user-images.githubusercontent.com/214002/64908093-bac32f80-d6fb-11e9-8bdd-d788c9d33b82.png)
I’d expect only `OGH 2 Ob 328/97t.` in the footnote. The reporters should only be added if there were any locators.

georgd replied:
So, I found out that in Word I used the wrong citations. What I get now is almost as expected in the bibliography:
![image](https://user-images.githubusercontent.com/214002/64946053-4b238080-d872-11e9-8c99-89fa396425a1.png)
Only the delimiters aren’t quite right. There is `., = ` between the parallel citations. I have no idea, where the comma comes from, the period is most likely the layout suffix.

For the citation, there’s probably a conceptual issue. I need only the common part for all parallel citations if there’s no locator, so only the first one prints anything. But I need all of them in the bibliography.

georgd committed:
Add missing citation items

georgd commented:
Now, I’m getting an error:
```
TypeError: Cannot read property '0' of undefined
      at Sys.run (/usr/local/lib/node_modules/citeproc-test-runner/lib/sys.js:286:28)
      at Context.<anonymous> (.cslTestFixtures/fixtures.js:876:27)
      at processImmediate (internal/timers.js:439:21)
```

georgd committed:
Correct seeAlso relations

georgd committed:
Only test bibliography

georgd committed:
Correct title sub delimiter

georgd commented:
@fbennett I’m still having troubles with parallel citations in test028. I have now changed the test to only build the bibliography entry and the error is:

```
1) Integration tests
       AT Case with parallel citations style_test028:
     SyntaxError: Unexpected token u in JSON at position 0
      at JSON.parse (<anonymous>)
      at CSL.Engine.retrieveItem (/usr/local/lib/node_modules/citeproc-test-runner/node_modules/citeproc/citeproc_commonjs.js:4161:17)
      at CSL.Engine.makeCitationCluster (/usr/local/lib/node_modules/citeproc-test-runner/node_modules/citeproc/citeproc_commonjs.js:7681:21)
      at Sys.run (/usr/local/lib/node_modules/citeproc-test-runner/lib/sys.js:373:43)
      at Context.<anonymous> (/Users/gduffner/.cslTestFixtures/fixtures.js:930:27)
      at processImmediate (internal/timers.js:458:21)
```

The style looks like this (legal case only, extracted from style and style-module):

```xml
<macro name="juris-title">
  <text variable="title" form="short" prefix="[" suffix="]"/>
</macro>
<macro name="juris-main">
  <group delimiter=" " parallel-first="authority genre number issued" parallel-delimiter-override=" = ">
    <group no-repeat="authority">
      <names variable="authority">
        <name/>
        <institution institution-parts="short">
          <institution-part name="short"/>
        </institution>
      </names>
    </group>
    <text variable="genre" prefix="(" suffix=")"/>
    <group delimiter=", ">
      <choose>
        <if>
          <conditions match="any">
            <condition variable="container-title" match="none"/>
            <condition context="bibliography"/>
          </conditions>
          <date variable="issued" form="numeric"/>
        </if>
      </choose>
      <number variable="number"/>
    </group>
  </group>
</macro>
<macro name="juris-tail">
  <group delimiter=" ">
    <text variable="container-title" prefix="= "/>
    <group delimiter=", ">
      <group delimiter="/">
        <number variable="volume"/>
        <date variable="publication-date" date-parts="year" form="numeric"/>
        <number variable="collection-number"/>
        <number variable="issue"/>
      </group>
      <number variable="page"/>
    </group>
    <names variable="commenter" font-style="italic" prefix="(" suffix=")">
      <name/>
    </names>
    <text variable="references" prefix="= "/>
  </group>
</macro>
<bibliography>
  <layout suffix=".">
    <group delimiter=" ">
      <group delimiter=" ">
        <text macro="juris-main"/>
        <text macro="juris-title"/>
        <text macro="juris-tail"/>
      </group>
    </group>
  </layout>
</bibliography>
```

georgd commented:
Closing this PR; In the meantime a new edition of the style guide was published so I‘m going to commit some fundamental changes.

georgd commented:
> @fbennett I’m still having troubles with parallel citations in test028. I have now changed the test to only build the bibliography entry and the error is:

...

I’ve finally found my syntax mistakes and was able to put up a new test file that shows this issue where parallel citations work in citations but not in bibliographies. Posted it in https://github.com/Juris-M/citeproc-js/issues/148

fbennett commented:
Thanks. I'll be able to look into this after the work on jurisdiction data is done.
